### DISCUSSION TITLE
OTA update

### DISCUSSION THREAD

**Discussion by mauroForlimpopoli** [Moddable-OpenSource/moddable#611] [Ideas]
I will partition my memory in Ota_0 e Ota_1. If I have understand right, before the OTA the program run from Ota_0 and after the OTA, the program will start from Ota_1, if I do a new OTA the program starts from Ota_0...and so on. Can I split the memory in different size of OTA? In other hands, I wish to set Ota_0 to 300kB, and Ota_1 to 3MB and when the device starts, do an automatic update of (up*bin). In the new program there is a manual request of OTA. When another OTA is request, boot*bin will be charged and it will do another automatic OTA of up.bin. So I can have more space for the main program. Is it possible ? Parts of memory partition is defined in the manifest (stack...static...keys). Is there a way to change them dinamically ?

**Comment by phoddie** [comment]
In theory what you describe is possible.

> Parts of memory partition is defined in the manifest (stack...static...keys). Is there a way to change them dynamically ?

That's impractical, at best. But why would you need to do that? You can build one firmware image for the small partition and a another one for the larger partition. They are entirely separate firmware builds so you can configure each separately.

**Reply by mauroForlimpopoli** [reply]
I did that andI have some questions : 
this is my main code :
/*
 * Copyright (c) 2016-2017  Moddable Tech, Inc.
 *
 *   This file is part of the Moddable SDK.
 * 
 *   This work is licensed under the
 *       Creative Commons Attribution 4.0 International License.
 *   To view a copy of this license, visit
 *       <https://creativecommons.org/licenses/by/4.0>.
 *   or send a letter to Creative Commons, PO Box 1866,
 *   Mountain View, CA 94042, USA.
 *
 */

import SNTP from "sntp";
import WiFi from "wifi";
import { Request } from "http";

import OTA from "esp32/ota";

trace("ok"+ "\n"	);

This is the manifest:
{	
	"build": {
		"PARTITIONS_FILE": "./partitions.csv"
	},
	"include": [
		"$(MODDABLE)/examples/manifest_base.json",
		"$(MODDABLE)/examples/manifest_piu.json",
		"$(MODULES)/network/wifi/manifest.json",
		"$(MODDABLE)/examples/manifest_net.json",
		"$(MODULES)/pins/i2c/manifest.json"
	],
	"modules": {
		"*": [
			"./main",
			"$(MODULES)/network/http/*"
		],
		"esp32/ota": "$(MODDABLE)/build/devices/esp32/modules/ota/*"
	},
	"preload": [
		"http",
		"esp32/ota"
	]
}

this the pertition:

nvs,		data,	nvs,		0x9000,			0x004000,
otadata,    data,   ota,    	0x00d000,   	0x2000,
phy_init,	data,	phy,		0xf000,			0x001000,
ota_0,    	app, 	ota_0,   	0x010000,		0x100000,
ota_1,    	app,  	ota_1,   	0x110000, 		0x190000,
storage,	data,	spiffs,	0x300000,		512K,
xs,		0x40,	1,		0x3D0000,		0x010000,
settings,	data,	1,		0x3E0000,		0x010000

1) No error, bur the program doesn't start.
2) The "xs_esp32.bin" created if about 1MB. Is there a way to reduce it ?
in the "https://blog.moddable.com/blog/clock-project/" I seen that maybe I need to reset the device with 
"esptool.py --port /dev/cu.SLAB_USBtoUART erase_region 0xd000 0x2000",
Where must I execute that ? Inside my "C:\esp32\esp-idf\components\esptool_py\esptool" always I have error .....

**Reply by phoddie** [reply]
The partition CSV data doesn't look correct. I would start by getting that resolved. Here's the line for OTA 1:

```
ota_1, app, ota_1, 0x110000, 0x190000,
```

The next partition begins at 0x2A0000 (0x110000 + 0x190000). The next partition specified is storage with begins at 0x300000, leaving 0x060000 bytes unaccounted for. Also, the storage line lists the length as "512K" where as all the other lines used 0x0.... notation. I'm not sure if the KB notation is supported. Assuming it is, the xs section should begin at 0x300000 + 512K (0x80000) == 0x380000, not 0x3D0000  as shown.

**Reply by mauroForlimpopoli** [reply]
I changed the partition.csv. these are the new partitions:
nvs,		data,	nvs,		0x9000,			0x004000,
otadata,    data,       ota,    	0x00d000,   	        0x2000,
phy_init,	data,	phy,		0xf000,			0x001000,
ota_0,    	app, 	ota_0,   	0x010000,		0x200000,
ota_1,    	app,  	ota_1,   	0x210000, 		0x0F0000,
storage,	data,	spiffs,	0x300000,		0x0D0000,
xs,		0x40,	1,		0x3D0000,		0x010000,
settings,	data,	1,		0x3E0000,		0x010000

this is the code of main for OTA:
let ota = new OTA;
function updateCode () {
trace("up")
	let request = new Request({host: "www.madesistemi.it", path: "/ota/firmware.bin"});
	request.callback = function(message, value, etc) {
trace(message+"  m\n")
		if (4 == message) {
			let bytes = this.read(ArrayBuffer);
			trace(bytes.byteLength, "\n");
			ota.write(bytes);
		}
	
		if (5 == message) {
			trace(`\n\nTransfer complete.\n`);
			ota.complete();
		}
	}
};

when the funtion is called these are che trace
up1  m
2  m
2  m
2  m
2  m
2  m
2  m
2  m
2  m
3  m
4  m
166
C:\Users\fly3\projects\moddable\build\devices\esp32\modules\ota\ota.c (77) # Break: (host): write failed!

**Reply by phoddie** [reply]
Progress! That error is coming from the ESP IDF API `esp_ota_write`. It was able to create the OTA session -- `esp_ota_begin` must have succeeded earlier.  You could try tracing the error from `esp_ota_write` to see if it gives a clue. Add a call to `xsLog` immediate after `esp_ota_write`. It should trace the error number to the console.

```c
	err = esp_ota_write(ota->handle, data, size);
	xsLog("esp_ota_write err=%d\n", (int)err);
```

The decoder ring for the errors codes is [here](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/error-codes.html).

**Reply by mauroForlimpopoli** [reply]
This is the error:
ESP_ERR_OTA_VALIDATE_FAILED (0x1503): Error if OTA app image is invalid.
the file "firmware.bin" is the file "xs_esp32.bin" renamed from my directory 
C:\Users\fly3\projects\moddable\build\bin\esp32\moddable_two\debug\ota

**Reply by mauroForlimpopoli** [reply]
I tried with other file but I get always the same error
...I think the error means "First byte of image contains invalid app image magic byte."

**Reply by phoddie** [reply]
So, what is the first byte? It could be that the server is returning an error message. Instead of reading an ArrayBuffer try a string. Then you can trace it:

trace(this.read(String), "\n");

**Reply by phoddie** [reply]
This:

```
curl http://www.madesistemi.it/ota/firmware.bin
```

Returns the following:

```
<head><title>Document Moved</title></head>
<body><h1>Object Moved</h1>This document may be found <a HREF="https://www.madesistemi.it/ota/firmware.bin">here</a></body> 
```

The server is redirecting the http request to https. The client doesn't expect that. There are two options: change the server to allow http requests to work or change the client to make an https request instead.

**Reply by mauroForlimpopoli** [reply]
I linked to another host where It is not present the redirect to https and now it works.
Your assistance is always on time and correct.
Thank you
